File: leetcode/go/construct_binary_tree_from_inorder_and_postorder_traversal.go
/*
 * Given inorder and postorder traversal of a tree, construct the binary tree.
 *
 * Note:
 * You may assume that duplicates do not exist in the tree.
 *
 * For example, given
 *
 * inorder = [9,3,15,20,7]
 * postorder = [9,15,7,20,3]
 * Return the following binary tree:
 *
 *     3
 *    / \
 *   9  20
 *     /  \
 *    15   7
 */

package main

func buildTreeUtil(inorder []int, postorder []int, index *int) *TreeNode {
	if len(inorder) == 0 {
		(*index)++
		return nil
	}

	root := &TreeNode{Val: postorder[*index]}

	find := func(arr []int, key int) int {
		for i := range arr {
			if arr[i] == key {
				return i
			}
		}
		return -1
	}

	inoIndex := find(inorder, postorder[*index])

	(*index)--
	root.Right = buildTreeUtil(inorder[inoIndex+1:], postorder, index)
	(*index)--
	root.Left = buildTreeUtil(inorder[:inoIndex], postorder, index)

	return root
}

func buildTree(inorder []int, postorder []int) *TreeNode {
	index := len(postorder) - 1
	return buildTreeUtil(inorder, postorder, &index)
}

func main() {
	tests := []struct {
		inorder   []int
		postorder []int
	}{
		{
			[]int{},
			[]int{},
		},
		{
			[]int{9},
			[]int{9},
		},
		{
			[]int{9, 3, 15, 20, 7},
			[]int{9, 15, 7, 20, 3},
		},
		{
			[]int{15, 9, 3, 20, 7},
			[]int{15, 9, 7, 20, 3},
		},
	}

	for _, t := range tests {
		root := buildTree(t.inorder, t.postorder)
		PrintTree(root)
	}
}

File: leetcode/python3/construct_binary_tree_from_inorder_and_postorder_traversal.py
"""
Given inorder and postorder traversal of a tree, construct the binary tree.

Note:
You may assume that duplicates do not exist in the tree.

For example, given

inorder = [9,3,15,20,7]
postorder = [9,15,7,20,3]
Return the following binary tree:

    3
   / \
  9  20
    /  \
   15   7
"""

import unittest
from typing import List
from tree_utils import TreeNode, PrintTree


class Solution:
    def __buildTree(self, inorder: List[int], postorder: List[int], index: int = -1) -> (TreeNode, int):
        print("postorder:", postorder, "  inorder:", inorder, "   index:", index)
        if not inorder:
            return None, index + 1

        root: TreeNode = TreeNode(val=postorder[index])
        ino_index = inorder.index(postorder[index])

        print("index:", index, "root:", root.val, "ino_index:", ino_index)

        root.right, index = self.__buildTree(inorder[ino_index + 1:], postorder, index - 1)
        root.left, index = self.__buildTree(inorder[:ino_index], postorder, index - 1)

        return root, index

    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        root, _ = self.__buildTree(inorder, postorder)
        return root


class TestSolution(unittest.TestCase):
    def test_buildTree(self):
        tests = [
            ([9, 3, 15, 20, 7], [9, 15, 7, 20, 3]),
            ([9], [9]),
            ([15, 9, 3, 20, 7], [15, 9, 7, 20, 3]),
        ]

        for t in tests:
            root = Solution().buildTree(*t)
            PrintTree(root)


if __name__ == "__main__":
    unittest.main()
